rules/REQUEST-944-APPLICATION-ATTACK-JAVA.conf

```diff
@@ -0,0 +1,210 @@
+SecRule REQUEST_BODY "@rx (\xac\xed\x00\x05)" \
+    "msg:'Magic bytes Detected, probable java serialization in use',\
+    phase:2,\
+    rev:'1',\
+    ver:'OWASP_CRS/3.1.0',\
+    pass,\
+    id:944300,\
+    tag:'application-multi',\
+    tag:'language-java',\
+    tag:'platform-multi',\
+    tag:'attack-rce',\
+    tag:'OWASP_CRS/WEB_ATTACK/COMMAND_INJECTION',\
+    tag:'WASCTC/WASC-31',\
+    tag:'OWASP_TOP_10/A1',\
+    tag:'PCI/6.5.2',\
+    tag:'paranoia-level/3',\
+    severity:'NOTICE',\
+    setvar:'tx.msg=%{rule.msg}',\
+    setvar:tx.rce_score=+%{tx.notice_anomaly_score},\
+    setvar:tx.anomaly_score=+%{tx.notice_anomaly_score},\
+    setvar:tx.%{rule.id}-OWASP_CRS/WEB_ATTACK/RCE-%{matched_var_name}=%{tx.0},\
+    setvar:tx.serialized_object=1"
+
+# Looking for possibly RCE vulnerable java classes if magic bytes were detected,
+# anomaly score set to warning as both conditions may indicate a possibly vulnerable application
+SecRule &TX:serialized_object "!@eq 0" \
+    "msg:'Probable java serialization containing vulnerable java class in use',\
+    phase:2,\
+    rev:'1',\
+    ver:'OWASP_CRS/3.1.0',\
+    block,\
+    id:944305,\
+    tag:'application-multi',\
+    tag:'language-java',\
+    tag:'platform-multi',\
+    tag:'attack-rce',\
+    tag:'OWASP_CRS/WEB_ATTACK/COMMAND_INJECTION',\
+    tag:'WASCTC/WASC-31',\
+    tag:'OWASP_TOP_10/A1',\
+    tag:'PCI/6.5.2',\
+    tag:'paranoia-level/3',\
+    severity:'WARNING',\
+    chain"
+        SecRule ARGS|XML|REQUEST_BODY "@rx (CloneTransformer|ForClosure|InstantiateFactory|InstantiateTransformer|InvokerTransformer|PrototypeCloneFactory|PrototypeSerializationFactory|WhileClosure)" \
+            "t:none,\
+            setvar:'tx.msg=%{rule.msg}',\
+            setvar:tx.rce_score=+%{tx.warning_anomaly_score},\
+            setvar:tx.anomaly_score=+%{tx.warning_anomaly_score},\
+            setvar:tx.%{rule.id}-OWASP_CRS/WEB_ATTACK/RCE-%{matched_var_name}=%{tx.0},\
+            setvar:tx.serialized_object_vulnerable_class=1"
+
+# Magic bytes detected and payload included possibly RCE vulnerable classess detected and process execution methods detected
+# anomaly score set to critical as all conditions indicate the request try to perform RCE.
+SecRule &TX:serialized_object_vulnerable_class "!@eq 0" \
+    "msg:'Remote Command Execution: Java serialization (CVE-2015-5842)',\
+    phase:2,\
+    rev:'1',\
+    ver:'OWASP_CRS/3.1.0',\
+    block,\
+    id:944310,\
+    tag:'application-multi',\
+    tag:'language-java',\
+    tag:'platform-multi',\
+    tag:'attack-rce',\
+    tag:'OWASP_CRS/WEB_ATTACK/COMMAND_INJECTION',\
+    tag:'WASCTC/WASC-31',\
+    tag:'OWASP_TOP_10/A1',\
+    tag:'PCI/6.5.2',\
+    tag:'paranoia-level/3',\
+    logdata:'Matched Data: %{TX.0} found within %{MATCHED_VAR_NAME}: %{MATCHED_VAR}',\
+    severity:'CRITICAL',\
+    chain"
+        SecRule ARGS|XML|REQUEST_BODY "(runtime|processbuilder)" \
```

Comment:
Can we use @rx and avoid captures here?

Reply:
no captures are required as you see no capture action is listed but just missed the ?: here and there

Reply:
944305 and 944310 uses vars to trigger, could be chained but in a variant of these I used the serialized_object var for both to detect attack variations or turn 944310 in something super generic as many other rules in the CRS as 944200 and use the send rule to differentiate possible interesting stuff from a very likely attack, the rules combined in a chain will solely trigger upon CVE identification but will not alert on the preconditions for other attacks.
Also keeping the rules independent can be helpful to add detection of more methods or classes that may be abused and not only a specific attack.
The payloads I've seen for both serialization attacks are noisy there are lots of special characters, and a few other rules will get on the way especially on PL4 but as you know it is perfectly normal to include a single rule file and not the entire rule/*.conf especially if you have no php, java or whatever.